tests/stdlib/tos.nim

```diff
@@ -333,3 +365,108 @@ block ospaths:
   doAssert joinPath("", "lib") == "lib"
   doAssert joinPath("", "/lib") == unixToNativePath"/lib"
   doAssert joinPath("usr/", "/lib") == unixToNativePath"usr/lib"
+
+block parentDir:
+  let examples = [
+    ("/usr/local/bin", "/usr/local"),
+    ("foo/bar.nim", "foo"),
+    ("foo//bar.nim", "foo"),
+    ("foo//bar//", "foo"),
+    ("foo///bar", "foo"),
+    ("foo///bar/.", "foo"),
+    ("./foo///bar", "./foo"),
+    (".//foo///bar", ".//foo"),
+    ("/.//foo///bar", "/.//foo"),
+    ("foo/bar//a/./.", "foo/bar"),
+    ("/a/bar", "/a"),
+    ("/bar", "/"),
+
+    # same as in shell, `cd ..` returns `.` when pwd = "foo"; "." is not same as empty
+    # see https://github.com/nim-lang/Nim/pull/10018#issuecomment-447996816
+    ("a/./.", "."),
+    ("./bar", "."),
+    (".//bar", "."),
+    ("bar", "."),
+
+    (".git", "."),
+    (".git.bak1", "."),
+
+    # with parentDirOfRootIsEmpty=false
+    # ("/", "/"),
+    # ("/.", "/"),
+    # ("/..", "/"),
+    # ("/./", "/"),
+    # fix #8734 (bug 3)
+    # ("/", "/"),
+
+    # with parentDirOfRootIsEmpty=true
+    ("/", ""),
+    ("/.", ""),
+    ("/..", ""),
+    ("/./", ""),
+
+    # return empty when no parent possible
+    ("", ""),
+    (".", ""),
+    ("./", ""),
+    ("..", ""),
+    ("../", ""),
+    ("../..", ""),
+
+    # regression tests
+
+    # fix #8734 (bug 2)
+    ("a/b//", "a"),
+    ("a/b/", "a"),
+
+
+    # fix #8734 (bug 4)
+    ("/a.txt", "/"),
+  ]
+
+  doAssert runTestCases("parentDir", examples, parentDir(it))
```

Comment:
That is cargo-cult programming. Now I can look at a list of tuples all day long before I can see it's about ``parentDir``. DRY and meta programming are for creating higher abstractions, not to save keystrokes.

Reply:
> Now I can look at a list of tuples all day long before I can see it's about `parentDir`

I added documentation to make everything clear. 

I had initially started with the `one doAssert per test case` but that became quickly un-manageable with the number of test cases, in particular:
* the important part of each test case was lost in noise of the repetitive surrounding code, eg `  doAssert parentDir("/usr/local/bin".unixToNativePath) == "/usr/local".unixToNativePath` ; instead, I'm only highlighting the part that changes
* on failure, it doesn't tell you what `parentDir` produced as output (because you don't like `assertEquals` for some reason)
* it stops on 1st error instead of showing all errors, making it terrible when you need to change the implementation of `parentDirs`; instead I'm showing all errors before failing
* if you have to change something, you have to change it everywhere; unlike here where I need to change it in only one place. yes, it's `DRY`, it's not `cargo cult`

## note: the lack of this kind of thorough testing as done here for `parentDir` is the root cause for all those issues and regressions I keep running into when using `std/os` procs (eg https://github.com/nim-lang/Nim/pull/10018 , https://github.com/nim-lang/Nim/pull/10047 , etc)

I have other PR's in the pipeline that fix other bugs in std/os (eg `"/a" /../ "d"`  should be `"/d"`, not `"d"` + other bugs with other procs) so let's please move on with this PR

Reply:
> the lack of this kind of thorough testing as done here for `parentDir` is the root cause for all those issues and regressions I keep running into when using `std/os` procs (eg #10018 , #10047 , etc)

Maybe but it's just you. Most people don't use double slashes in all their paths. And your improved implementation adds plenty of its own, new problems.

Reply:
@araq
* I don't understand your comment. #10018, #10047, #10545 and other PRs I wrote for path handling fix bugs and regressions unrelated to double slashes; incorrect handling of paths with double slashes were just one of the issues I've fixed.
* all languages I've worked with correctly handle paths with double slashes, eg: C++(boost), D, python, java, js(node) etc...; as I already explained
* Double slashes can occur from a variety of reasons, eg:
```
export buildDir=/path/to/build/
export logFile=$buildDir/log.txt # => expands to a path with double slash
```
* see also https://unix.stackexchange.com/questions/212805/should-i-use-a-slash-at-the-end-of-path-variables-in-shell-script-or-not 
> To be on the safe side, include the slash. This can lead to multiple slashes when concatenating the paths, but at least you avoid problems.

and this: https://unix.stackexchange.com/questions/1910/how-does-linux-handle-multiple-consecutive-path-separators-home-username
> Multiple slashes are allowed and are equivalent to a single slash

> And your improved implementation adds plenty of its own, new problems

I'm not sure which are these new problems you're referring to, after this PR changed from `parentDir("/") = "/"` to `""`; which code was previously correct that became incorrect after this PR?